### DISCUSSION TITLE
SOPS with Azure Key Vault

### DISCUSSION THREAD

**Discussion by haraldatbmw** [fluxcd/flux2#424] [Q&A] [answered]
I try to use SOPS for secret encryption with Azure AKS. Encryption with SOPS CLI works well. 
But decryption in the kustomize-controller does not work.

The kustomize-controller logs this error:

```
decryption failed for 'feedy-nexus': GetDataKey: Error getting data key: 0 successful groups required, got 0
```

The flux guide says:
```
When using AWS/GCP KMS or Azure Key Vault, you'll have to bind an IAM Role with read access to the KMS keys to the default service account of the flux-system namespace for kustomize-controller to be able to fetch keys from KMS.
```

Can anybody describe how I grant read access to Azure key vault from the flux-system default service account?
It's not clear to me which steps are needed.

**Comment by phillebaba** [comment, marked as answer]
So to access Key Vault you need to authenticate towards Azure both when encrypting and decrypting. You have a a couple authentication options when using the cli. My guess is you have configured this locally when using the cli.
https://github.com/mozilla/sops#encrypting-using-azure-key-vault

For this to work in the Kustomization controller you either need to configure a Service Principal with environment variables in the container or use MSI through aad-pod-identity.
https://github.com/Azure/aad-pod-identity

**Reply by haraldatbmw** [reply]
@phillebaba If I configure service principal via env-vars on the kustomize-controller I need to put the service principal secret in my git repo.
And this is what I want to prevent using SOPS. So it seams not to be a valid way.

- AZURE_TENANT_ID | The ID for the Active Directory tenant that the service principal belongs to.
- AZURE_CLIENT_ID | The name or ID of the service principal.
- AZURE_CLIENT_SECRET | The secret associated with the service principal.

**Reply by stefanprodan** [reply]
You can load the [env vars from a Kubernetes secret](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables). I would create the Kubernetes secret at cluster creation time without storing it in Git. With the Terraform Kubernetes provider, you can create the flux-system namespaces and the Kubernetes secret before deploying Flux.

**Reply by phillebaba** [reply]
@haraldatbmw if you want something quick go with Stefans option.

I personally prefer using aad-pod-identity as you no longer have to deal client secrets for pods. It is a bit more work, but in my opinion a bigger payoff in the end. When you use aad-pod-identity all you need to do is add an extra label to the kustomize controller, no secrets needed in your git repository.

**Reply by stefanprodan** [reply]
@phillebaba can you please add a link to aad-pod-identity in the SOPS docs, I assumed AKS lets you bind an IAM Role to a service account like EKS/GKE...

**Reply by phillebaba** [reply]
Sure, I created an issue for it #427. Can you assign it to me?

**Reply by haraldatbmw** [reply]
I was able to install an configure **aad-pod-identity** on my cluster and use **SOPS** with `AZURE_AUTH_METHOD=msi` inside the kustomize-controller for decryption of my secrets. 

I found a really helpful article on how to install aad-pod-identity here:
https://dev.to/mpuhacz/the-right-way-of-accessing-azure-services-from-inside-your-azure-kubernetes-cluster-4p8d

For my current small projects the whole architecture with SOPS, aad-pod-identity, Azure Key Vault and Azure Managed Identity seams really oversized to manage secret encryption only. And because it uses a lot of azure specific components it's definitely not cloud agnostic. So I will prefer a solution with bitnami's sealed-secrets.

**Reply by stefanprodan** [reply]
What made you chose sealed-secrets over SOPS with PGP?

**Reply by haraldatbmw** [reply]
Ohh, I totally skipped that in my mind and thought I have to use SOPS with the key-management solution of my cloud provider.
But when using SOPS with PGP it's even simpler in architecture than sealed-secrets because no additional component have to be installed. Only the PGP key. Am I right?

**Reply by stefanprodan** [reply]
Yes and not only that, but SOPS CLI has some really nice helpers like rotating keys and updating secrets in Git.

**Reply by stefanprodan** [reply]
I would generate a PGP key for Flux, store that in KMS for disaster recovery, then use SOPS to encrypt the `data` field of every secret in the gitops repo.

**Reply by haraldatbmw** [reply]
@stefanprodan Thanks for clarifying this!

**Comment by stealthybox** [comment]
Some docs for doing this with AAD-Pod-Identity and Service Principal Secrets is included with #1064

**Reply by tz-torchai** [reply]
Hi Leigh! Thanks for adding the docs for Azure KeyVault!

I am using flux + sops + Hashicorp Vault by following [this guide](https://github.com/mozilla/sops#encrypting-using-hashicorp-vault).

I am really curious that how can i let flux know which vault address it's going to call to decrypt the data? and how can it authenticate with Vault server?

Thanks in advance!
